src/core/SampleBufferV2.cpp

```diff
@@ -0,0 +1,220 @@
+	{
+		loadFromBase64(strData);
+	}
+}
+
+SampleBufferV2::SampleBufferV2(const sampleFrame* data, const int numFrames)
+	: m_sampleData(data, data + numFrames)
+	, m_filePath("")
+	, m_originalSampleRate(Engine::audioEngine()->processingSampleRate())
+{
+}
+
+SampleBufferV2::SampleBufferV2(const int numFrames)
+	: m_sampleData(numFrames)
+	, m_filePath("")
+	, m_originalSampleRate(Engine::audioEngine()->processingSampleRate())
+{
+}
+
+SampleBufferV2::SampleBufferV2(SampleBufferV2&& other)
+	: m_sampleData(std::move(other.m_sampleData))
+	, m_filePath(std::exchange(other.m_filePath, std::nullopt))
+	, m_originalSampleRate(std::exchange(other.m_originalSampleRate, 0))
+{
+	other.m_sampleData.clear();
+}
+
+SampleBufferV2& SampleBufferV2::operator=(SampleBufferV2&& other)
+{
+	if (this == &other) { return *this; }
+
+	m_sampleData = std::move(other.m_sampleData);
+	m_filePath = std::exchange(other.m_filePath, std::nullopt);
+	m_originalSampleRate = std::exchange(other.m_originalSampleRate, 0);
+	other.m_sampleData.clear();
+
+	return *this;
+}
+
+const std::vector<sampleFrame>& SampleBufferV2::sampleData() const
+{
+	return m_sampleData;
+}
+
+const std::optional<std::filesystem::path>& SampleBufferV2::filePath() const
+{
+	return m_filePath;
+}
+
+int SampleBufferV2::originalSampleRate() const
+{
+	return m_originalSampleRate;
+}
+
+std::string SampleBufferV2::toBase64() const
+{
+	const char* rawData = reinterpret_cast<const char*>(m_sampleData.data());
+	QByteArray data = QByteArray(rawData, m_sampleData.size() * sizeof(sampleFrame));
+	return data.toBase64().constData();
+}
+
+int SampleBufferV2::numFrames() const
+{
+	return m_sampleData.size();
+}
+
+void SampleBufferV2::resample(const int oldSampleRate, const int newSampleRate)
+{
+	int dstFrames = static_cast<int>(static_cast<float>(numFrames()) / oldSampleRate * newSampleRate);
+	auto resampleBuf = std::vector<sampleFrame>(dstFrames);
+
+	int error;
+	SRC_STATE* state;
+	if ((state = src_new(SRC_LINEAR, DEFAULT_CHANNELS, &error)) != nullptr)
+	{
+		SRC_DATA srcData;
+		srcData.data_in = m_sampleData.data()->data();
+		srcData.input_frames = numFrames();
+		srcData.data_out = resampleBuf.data()->data();
+		srcData.output_frames = dstFrames;
+		srcData.src_ratio = static_cast<double>(newSampleRate) / oldSampleRate;
+		srcData.end_of_input = 1;
+
+		error = src_process(state, &srcData);
+		src_delete(state);
+	}
+
+	if (error != 0)
+	{
+		throw std::runtime_error(std::string("An error occurred when resampling: ") + src_strerror(error) + '\n');
+	}
+
+	m_sampleData = std::move(resampleBuf);
+}
+
+void SampleBufferV2::loadFromAudioFile(const std::filesystem::path& audioFilePath)
+{
+	SF_INFO sfInfo;
+	sfInfo.format = 0;
+
+	auto sndFileDeleter = [](SNDFILE* ptr) { sf_close(ptr); };
+	#ifdef LMMS_BUILD_WIN32
+		auto sndFile = std::unique_ptr<SNDFILE, decltype(sndFileDeleter)>(
+			sf_wchar_open(audioFilePath.c_str(), SFM_READ, &sfInfo), sndFileDeleter);
```

Comment:
I forgot to mention that you have to define `ENABLE_SNDFILE_WINDOWS_PROTOTYPES` as 1 before including `sndfile.h` prior to libsndfile 1.1.0, in order to use `sf_wchar_open`.

Reply:
Currently, some ogg files crash which Saker says has something to do with sndfile. Does this one fix it?

Reply:
> I forgot to mention that you have to define `ENABLE_SNDFILE_WINDOWS_PROTOTYPES` as 1 before including `sndfile.h` prior to libsndfile 1.1.0, in order to use `sf_wchar_open`.

Since the plan is to use FFmpeg instead of sndfile, I'll just make the switch once it's added into the codebase.

Reply:
Just as a follow up to this: The plan is still to use FFmpeg, but for now sndfile is still being used. I'll check if the problem still remains but my course of immediate action against this bug is unknown.

Reply:
Instead of ``sf_open`` and ``sf_wchar_open``, I went ahead and used sf_open_fd with ``QFile`` to simplify the handling of different file path formats on varying platforms, such as Windows.